torch/csrc/Device.cpp

```diff
@@ -169,6 +169,36 @@ PyObject* THPDevice_reduce(PyObject* _self, PyObject* noargs) {
   END_HANDLE_TH_ERRORS
 }
 
+PyObject* THPDevice_enter(PyObject* self, PyObject* noargs) {
+  HANDLE_TH_ERRORS
+  py::object mode = py::module::import("torch.utils._device")
+                        .attr("DeviceContext")(py::handle(self));
+  at::impl::PythonTorchFunctionTLS::push_onto_stack(
+      std::make_shared<c10::SafePyObject>(
+          mode.release().ptr(), getPyInterpreter()));
+  // So that with torch.device('cuda') as dev: works
+  Py_INCREF(self);
+  return self;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject* THPDevice_exit(PyObject* self, PyObject* unused) {
+  HANDLE_TH_ERRORS
+  at::impl::PythonTorchFunctionTLS::pop_stack();
```

Comment:
Should we assert that we get a DeviceContext here?

Reply:
TorchDispatchMode doesn't currently test for this:

```
    def __enter__(self):
        _push_mode(self)
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        _pop_mode()
```

I suppose we could add it but we ought to add it everywhere if we do.